### PR TITLE
add sql for geography

### PR DESCRIPTION
**Pull request by openrijal**

### CODE REVIEW

File: sql/geography.sql [hunk 1]
```diff
@@ -0,0 +1,971 @@
...
+Bhojpur,1236,2016,province,Bhojpur,district,1
+Dhankuta,1299,2016,province,Dhankuta,district,1
+Morang,1326,2016,province,Morang,district,1
+Sankhuwasabha,1394,2016,province,Sankhuwasabha,district,1
+Sunsari,1467,2016,province,Sunsari,district,1
+Terathum,1475,2016,province,Terathum,district,
+Ilam,1538,2016,province,Ilam,district,1
+Jhapa,1581,2016,province,Jhapa,district,1
+Panchthar,1658,2016,province,Panchthar,district,1
+Taplejung,1658,2016,province,Taplejung,district,1
```
Comment (cliftonmcintosh) [inline sql/geography.sql]: 1658 is a duplicate key here. There is a uniqueness constraint on `geo_code` and `geo_level`. I am not sure if there are other duplicates. When I tried to populate the table with this file, it errored here.

I used `cat geography.sql | psql -U wazimap_np -W wazimap_np` to try to populate the table using this file.

**Review by cliftonmcintosh — APPROVED** [review]

**Review by cliftonmcintosh — CHANGES_REQUESTED** [review]
records must conform to uniqueness constraint.

File: sql/geography.sql (hunk 1 above)
Comment (cliftonmcintosh) [reply sql/geography.sql]: (Accidentally hit "approve")

**Comment by openrijal** [comment]
The source oknp data has a lot of duplicates
`➜  nepalmap cut -d, -f2 geography_data.csv | sort | uniq -dc`


      2 1
      2 100
      2 1658
      2 190
      3 2
      2 244
      2 268
      2 3
      2 319
      2 4
      2 496
      2 5
      2 507
      2 51
      2 528
      2 545
      2 550
      2 6
      2 614
      3 630
      2 650
      3 671
      2 672
      2 676
      3 678
      2 679
      2 680
      2 7
      2 81

**Comment by openrijal** [comment]
I'm sure the reason is because each data begins with 1, for province, district and local, and if it increases sequentially then it is certain that they clash in those levels

I'll look into this tomorrow, may be come up with something.

**Comment by openrijal** [comment]
avoiding duplicates for integers has proved to be difficult than I expected

I see that geocode can be varchar(10), which means something like `pro1`, `dis6736` or `loc44352` should be fine

do you think we can use such codes? it would definitely avoid duplicates and get the work done.

cc: @cliftonmcintosh

**Comment by cliftonmcintosh** [comment]
@crackjack

I think that should be fine. I don't know of anything in the Wazimap framework that requires the geocode to be an integer.

**Comment by openrijal** [comment]
updated PR, please try populating the table now cc: @cliftonmcintosh

**Comment by openrijal** [comment]
new PR with geojson and topojson files matching geoids as in the csv for database entry

cc: @cliftonmcintosh

**Comment by cliftonmcintosh** [comment]
@crackjack 

Thanks for the update. Here are some discrepancies I've noticed.

1. Nawalparasi and Rupandehi each have two rows.

```
wazimap_np=> select * from wazimap_geography where geo_level = 'district' and name in ('Nawalparasi', 'Rupandehi');
 id | geo_level | geo_code |    name     | year | square_kms | parent_level | parent_code |  long_name  
----+-----------+----------+-------------+------+------------+--------------+-------------+-------------
 64 | district  | 3579     | Nawalparasi | 2016 |            | province     |             | Nawalparasi
 65 | district  | 3585     | Nawalparasi | 2016 |            | province     |             | Nawalparasi
 67 | district  | 3703     | Rupandehi   | 2016 |            | province     | pro-5       | Rupandehi
 68 | district  | 3715     | Rupandehi   | 2016 |            | province     | pro-5       | Rupandehi
(4 rows)
```

It looks like all of the local bodies for those districts have been mapped to the district rows whose geo_codes are "3585" and "3703". That means the rows for "3579" and "3715" can likely be removed, but we should make sure that the geojson and topojson files match.

2. The `parent_code` in the local rows does not match the `geo_code` for districts. The district `geo_code`s are integers, but the `parent_code`s for local bodies all start with "dis-". They have to match. Think of the `parent_code` as a foreign key that points to the `geo_code` of a row's parent.

For example, if Kathmandu district's `geo_code` is "100", then all of the localities in Kathmandu district should have a `parent_code` of "100", not "dis-100".

3. Some of the district rows are missing their parent ids. 

```
wazimap_np=> select * from wazimap_geography where geo_level = 'district' and parent_code not like 'pro%';;
 id | geo_level | geo_code |      name      | year | square_kms | parent_level | parent_code |   long_name    
----+-----------+----------+----------------+------+------------+--------------+-------------+----------------
 18 | district  | 1475     | Terathum       | 2016 |            | province     |             | Terathum
 25 | district  | 190      | Kavrepalanchok | 2016 |            | province     |             | Kavrepalanchok
 48 | district  | 2807     | Rukum          | 2016 |            | province     |             | Rukum
 58 | district  | 3212     | Manag          | 2016 |            | province     |             | Manag
 60 | district  | 3324     | Tanahun        | 2016 |            | province     |             | Tanahun
 63 | district  | 3487     | Kapilbastu     | 2016 |            | province     |             | Kapilbastu
 64 | district  | 3579     | Nawalparasi    | 2016 |            | province     |             | Nawalparasi
 65 | district  | 3585     | Nawalparasi    | 2016 |            | province     |             | Nawalparasi
 83 | district  | 81       | Sindhupalchowk | 2016 |            | province     |             | Sindhupalchowk
(9 rows)
```

The `parent_code`s in those rows are not nulls. They are also not a single empty space. I did not probe farther than that.

Is there something that is common about this that would explain how it happened?

**Comment by openrijal** [comment]
The districts are duplicate in the geojson files as well, its from the source that we use, will it be a problem if you keep the duplicates in the geojson but remove them from the CSV? @cliftonmcintosh

**Comment by cliftonmcintosh** [comment]
I am not sure what happens if there are duplicates in the geojson. It probably depends on how the map responds to a click.

**Comment by openrijal** [comment]
lets check that out, because I find it difficult to edit the geojson file and remove the duplicates, and without knowing which boundary is actual one

if we just pull the data from csv, that will only give us 1 item for Nawalparasi, that way we might be safe.

updated the PR just now, the main problem is that we don't have source of truth and gathering data from multiple source has caused all this problem

cc: @cliftonmcintosh

**Comment by cliftonmcintosh** [comment]
Right. We shouldn't have to edit the geojson files. They are too complex for that.

**Comment by openrijal** [comment]
I have created these new sql files from latest geojsons from Nikesh.

I still had to do a lot of manual work in the district geojson file, but I hope it is worth it.

There is also new geojson files, that might be needed to be replaced, I'll point you the source @cliftonmcintosh

**Comment by openrijal** [comment]
updated the geojson and topojson files in this PR

**Review by cliftonmcintosh — APPROVED** [review]
Thank you for all of your hard work getting these data imported.

**Pull request merged**